### PR TITLE
[Core][Scheme] Correcting & unifying implementation for InitializeNonLinearIteration

### PR DESCRIPTION
**Pull request by philbucher**
`scheme.h` calls the corresponding methods of elems/conds/constaints of `InitializeSolutionStep`, `FinalizeSolutionStep` and `FinalizeNonLinIteration`
The corresponding implementation for `InitializeNonLinIteration` was missing and is added in this PR
I have not run the tests yet since I want to first discuss if this is the correct change (I would say yes)

found by @KlausBSautter

### CODE REVIEW

**Review by loumalouomega — DISMISSED** [review]
I think I fixed this in some derived schemes, but I forgot to add on the base one. What about MPC?

File: kratos/solving_strategies/schemes/scheme.h [hunk 1]
```diff
@@ -502,6 +502,39 @@ class Scheme
...
+            auto it_cond = it_cond_begin + i;
+            it_cond->InitializeNonLinearIteration(r_current_process_info);
+        }
+
+        // Definition of the first constraint iterator
+        const auto it_const_begin = rModelPart.MasterSlaveConstraintsBegin();
+
+        // Initialize non-linear iteration for all of the constraints
+        #pragma omp parallel for
+        for(int i=0; i<static_cast<int>(rModelPart.MasterSlaveConstraints().size()); ++i) {
```
Comment (philbucher) [inline kratos/solving_strategies/schemes/scheme.h]: @loumalouomega

**Comment by loumalouomega** [comment]
OK, I haven't seen it. Go on with the merge ⛜

**Comment by loumalouomega** [comment]
What is missing to merge?

**Comment by philbucher** [comment]
> What is missing to merge?

Testing it more properly (aka running the tests)
At the time I did this I had no time and I wanted to put it up for discussion first

File: kratos/solving_strategies/schemes/scheme.h (hunk 1 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/schemes/scheme.h]: i think that this was deliberately not called here but rather directly within the builder and solver.

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
this is called element by element in the buildr and solver as i recall. take care not to call it twice

**Comment by philbucher** [comment]
> this is called element by element in the buildr and solver as i recall. take care not to call it twice

ok but isn't this inconsistent then?
Should we move those calls to the Scheme?

**Comment by RiccardoRossi** [comment]
I did it this way to allow for a usage of the type

~~~cpp
for(auto elem: model_part.Elements)
{
      elem.InitializeNonLinearIteration() //eventually here some memory could be allocated

      pScheme->CalculateSystemContributions(*(it.base()), LHS_Contribution, RHS_Contribution, EquationId, CurrentProcessInfo);

        Assemble(A, b, LHS_Contribution, RHS_Contribution, EquationId);

        pScheme->CleanMemory(*(it.base()));
}
~~~

so that an element could eventually allocate and deallocate something in the clean.
but i just checked the code and it looks like this behaviour is long gone

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
i'll remove my block. it is used as you are telling already in the staticscheme so the behaviour you are proposing is already standard

**Comment by philbucher** [comment]
@RiccardoRossi you are still blocking ...?

**Comment by RiccardoRossi** [comment]
yes i am. as far as i see we don t have a solution, and i rrally don t want a half backed kack to.go.official

**Comment by philbucher** [comment]
ok jut wanted to clarify

**Comment by RiccardoRossi** [comment]
sorry, i thought it is as a different PR. i can definitely remove my block on this

**Review by RiccardoRossi — DISMISSED** [review]

**Comment by philbucher** [comment]
@KratosMultiphysics/technical-committee in the scope of updating the `Scheme`s I took another look at this and we have to discuss the following things:
- **IN MASTER**: The `FinalizeSolutionStep` is called TWICE on Constraints, in the `Scheme` and in the `BuilderAndSolver`s!
=> I moved it to the `Scheme` such that it is consistent with `Element`s and `Condition`s
- The behavior of calling 
    - `InitializeSolutionStep`
    -  `FinalizeSolutionStep`
    - `InitializeNonLinIteration`
    - `FinalizeNonLinIteration`
   of `Element`s, `Condition`s and `Constraint`s was unified and is done in the `Scheme` (called from the `Strategy` => there is quite a mess going on in some derived `Scheme`s, hence I think this is really necessary
- What [@RiccardoRossi wrote above](https://github.com/KratosMultiphysics/Kratos/pull/5121#issuecomment-531109262) was still done in old `BuilderAndSolver`s, I removed those (since it is now done in the `Scheme`) and deprecated the corresponding functions of the `Scheme`
- I could not find where the `Constraint`s are initialized... This is neither done in the `Scheme` nor in the `BuilderAndSolver` not in the `Strategy`. @adityaghantasala @loumalouomega do you know?
- Some implementations that I am unifying here is done in a more or less complete manner in some derived `Scheme`s. I cleaned those right away otherwise I would again lose the overview
- In many `Scheme`s `InitializeNonLinearIteration` is called (form `Element`s and `Condition`s) before calculating system contributions, esp in the Fluid. Now that this is done in the BaseClass I think/hope that this can be safely removed. The `Scheme`s that are used for the structural problems already do this.

**SORRY THIS IS SUCH A BIG PR!** I would be happy to make smaller PRs! Here I wanted to collect all changes to make sure I don't miss anything!
I would very much appreciate if someone else could take a look through the code (not possible in Github) and check my implementation!

**Comment by philbucher** [comment]
I updated the Core Schemes and the ones in StructuralMechanics. I took a look at the ones in FluidDynamics, they are a HUGE mess :sweat_smile: 
=> TBD!

File: kratos/solving_strategies/schemes/scheme.h (hunk 1 above)
Comment (philbucher) [reply kratos/solving_strategies/schemes/scheme.h]: nope, the BuilderAndSolver never did this, I checked
Actually none did this I am quite sure

**Review by philbucher — COMMENTED** [review]
I just added the calls for the sake of completeness, I am not sure about the order!
=> Refers to the fluidschemes: https://github.com/KratosMultiphysics/Kratos/pull/5121/commits/6abf352a1ae663be9c72910f71cf97f651e68ce0

**Comment by philbucher** [comment]
ok nightly test also pass

**Comment by philbucher** [comment]
@KratosMultiphysics/technical-committee agrees about this change, but since it is a sensitive change @rubenzorrilla and I will take a close look soon to prevent things from breaking

FYI @sunethwarna @KratosMultiphysics/altair

**Comment by loumalouomega** [comment]
> @KratosMultiphysics/technical-committee agrees about this change, but since it is a sensitive change @rubenzorrilla and I will take a close look soon to prevent things from breaking
> 
> FYI @sunethwarna @KratosMultiphysics/altair

Good luck with the conflicts

**Review by loumalouomega — APPROVED** [review]
OK (if it passes the tests)

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Does not pass the tests

**Comment by rubenzorrilla** [comment]
@loumalouomega I think it's ready to be re-approved

**Review by loumalouomega — DISMISSED** [review]
Now

**Review by rubenzorrilla — DISMISSED** [review]
I carefully re-reviewed the files one-by-one and I think this can go in. However, I'm afraid that if any one has added a new scheme by copying & pasting and old one the missuses will still be there. Maybe it is a good idea that we merge the master and grep again for `InitializeNonLinearIteration`.

**Comment by philbucher** [comment]
I now went through the entire code and did the remaining updates, now we should be good to go :+1:

**Review by rubenzorrilla — APPROVED** [review]
I think that there are no more leftovers.

**Comment by philbucher** [comment]
merging before it causes further conflicts

This should not cause any problems, as esp in the fluid it was not deterministic how often `InitializeNonLinIteration` was called (e.g. when reactions are computed it was called too due to being called in CalcRHS)

**Pull request merged**